### PR TITLE
Support streaming mlsum dataset

### LINKED ISSUE #4572: Dataset Viewer issue for mlsum
### Link

https://huggingface.co/datasets/mlsum/viewer/de/train

### Description

There's seems to be a problem with the download / streaming of this dataset:

```
Server error
Status code:   400
Exception:     BadZipFile
Message:       File is not a zip file
```

### Owner

No

### PR DESCRIPTION
**Pull request by albertvillanova**
Support streaming mlsum dataset.

This PR:
- pins `fsspec` min version with fixed BlockSizeError: `fsspec[http]>=2021.11.1`
  - https://github.com/fsspec/filesystem_spec/pull/830
- unpins `s3fs==2021.08.1` to align it with `fsspec` requirement: `s3fs>=2021.11.1`
  > s3fs 2021.8.1 requires fsspec==2021.08.1
  - see discussion: https://github.com/huggingface/datasets/pull/2858/files#r700027326
- updates the following requirements to be compatible with the previous ones and one with each other:
  - `aiobotocore==1.4.2` to `aiobotocore>=2.0.1` (required by s3fs>=2021.11.1)
  - `boto3==1.17.106` to `boto3>=1.19.8` (to be compatible with aiobotocore>=2.0.1)
  - `botocore==1.20.106` to `botocore>=1.22.8` (to be compatible with aiobotocore and boto3)

Fix #4572.

### CODE REVIEW

File: src/datasets/download/streaming_download_manager.py [hunk 1]
```diff
@@ -406,8 +406,12 @@ def _prepare_http_url_kwargs(url: str, use_auth_token: Optional[Union[str, bool]
     # Fix Google Drive URL to avoid Virus scan warning
     if "drive.google.com" in url and "confirm=" not in url:
         url += "&confirm=t"
-    if url.startswith("https://raw.githubusercontent.com/"):
+    if (
+        url.startswith("https://raw.githubusercontent.com/")
         # Workaround for served data with gzip content-encoding: https://github.com/fsspec/filesystem_spec/issues/389
+        or url.startswith("https://gitlab.lip6.fr/")
```
Comment (lhoestq) [inline src/datasets/download/streaming_download_manager.py:412]: Why is this change needed ? On my side reading the file from this website in streaming mode works (though it takes some time to download)

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: It was not working on my side. I can recheck again...

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: @lhoestq if I remove this line of code, I get this error in streaming mode:
```
BlockSizeError: Got more bytes so far (>5255168) than requested (5251072)
```
I don't think their server supports HTTP Range requests and these must be disallowed.

Comment (lhoestq) [reply src/datasets/download/streaming_download_manager.py:412]: From the master branch, if I load the new MLSum dataset script (on https://github.com/huggingface/datasets/pull/4574/commits/48f1775c312955519c0a469096ae74a6c5841bde) it works fine
```python
ds = load_dataset("mlsum", "fr", revision="48f1775c312955519c0a469096ae74a6c5841bde", streaming=True)
next(iter(ds["train"]))
```
with fsspec==2022.3.0

What code do you run to get an error ?

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: @lhoestq from the main branch, loading exactly as you did above, I get an error:
```
BlockSizeError: Got more bytes so far (>5275016) than requested (5251072)
```
My fsspec version is 2021.08.01. I'm going to try with an updated version.

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: @lhoestq after updating `fsspec`, there is a dependency conflict:
- we require `s3fs==2021.08.1`
> s3fs 2021.8.1 requires fsspec==2021.08.1, but you have fsspec 2022.5.0 which is incompatible

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: Checked: with `fsspec==2022.5.0` there is no error and "mlsum" loads:
```
{'text': 'Jean-Jacques Schuhl, Gilles Leroy, Christian Gailly, Yasmina Khadra, James Ellroy, Amos Oz, V. S. Naipaul… Si la rentrée de janvier s\'annonce sous les meilleurs auspices, force est de constater que, avec ...
```

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: I'm investigating the requirement: `s3fs==2021.08.1`

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: @lhoestq I have found the discussion about pinning `s3fs`: https://github.com/huggingface/datasets/pull/2858/files#r700027326

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: I'm going to try to unpin it and see which version pip installs.

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py:412]: Also, I have checked which is the minimum version of `fsspec` that does not raise the error: `fsspec>=2021.11.1`. I'm pinning it.

Comment (albertvillanova) [reply src/datasets/download/streaming_download_manager.py]: When `s3fs` is unpinned, CI installs `s3fs-0.5.1`.

**Comment by albertvillanova** [comment]
After unpinning `s3fs` and pinning `fsspec[http]>=2021.11.1`, the CI installs
- `fsspec-2022.1.0`
- `s3fs-0.5.1`

and raises the following error:
```
 ImportError while loading conftest '/home/runner/work/datasets/datasets/tests/conftest.py'.
tests/conftest.py:13: in <module>
    import datasets
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/__init__.py:37: in <module>
    from .arrow_dataset import Dataset
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/arrow_dataset.py:62: in <module>
    from .arrow_reader import ArrowReader
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/arrow_reader.py:29: in <module>
    from .download.download_config import DownloadConfig
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/download/__init__.py:10: in <module>
    from .streaming_download_manager import StreamingDownloadManager
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/download/streaming_download_manager.py:20: in <module>
    from ..filesystems import COMPRESSION_FILESYSTEMS
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/filesystems/__init__.py:13: in <module>
    from .s3filesystem import S3FileSystem  # noqa: F401
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/datasets/filesystems/s3filesystem.py:1: in <module>
    import s3fs
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/s3fs/__init__.py:1: in <module>
    from .core import S3FileSystem, S3File
/opt/hostedtoolcache/Python/3.6.15/x64/lib/python3.6/site-packages/s3fs/core.py:12: in <module>
    from fsspec.asyn import AsyncFileSystem, sync, sync_wrapper, maybe_sync
E   ImportError: cannot import name 'maybe_sync'
```

I'm investigating this.

**Comment by lhoestq** [comment]
Maybe you can try setting the same minimum version as fsspec ? `s3fs>=2021.11.1`

**Comment by albertvillanova** [comment]
Yes, I have checked that they both require to have the same version. 

The issue then was coming from aiobotocore, boto3, botocore. I have changed them from strict to min version requirements.

**Comment by albertvillanova** [comment]
I have updated all min versions so that they are compatible one with each other. I'm pushing again...

**Comment by lhoestq** [comment]
Thanks !

**Comment by severo** [comment]
Nice!
